O3-614: Added support for AMPATH Forms Translations.

icrc-jofrancisco opened the pull request:
Support for loading AMPATH translation files using the initializer.
The file will be saved as `ClobDatatypeStorage` in order to support multiple forms.

rbuisson commented:
@icrc-jofrancisco , I guess the domain would be better named "Ampath Forms Translations"

gracepotma commented:
Perhaps @ibacher could review this?

mks-d reviewed (changes requested):
Thanks @icrc-jofrancisco, however we need a couple of things:
1. Unit tests.
2. Amendments to the README, including the line for the upcoming release notes, and this should probably be for 2.5.0 not 2.4.1, @rbuisson?
3. Possible addition of an ad-hoc README for this new domain, others to weigh in as to this is overkill or not.

ibacher commented:
> Possible addition of an ad-hoc README for this new domain, others to weigh in as to this is overkill or not.

I actually think documenting this domain is crucial.

icrc-jofrancisco commented:
Hi @ibacher @rbuisson @mks-d,
I added some tests, changed some names and completed the readme files.
I'm open to new suggestions or changes that might be needed.

Thanks,

mks-d reviewed (commented):
@icrc-jofrancisco did you see that the GitHub Actions build is failing? Could you have a look?

rbuisson commented:
@ibacher , didn't you mention that the AMPATH forms translation files should be attached to forms via a form resource?

rbuisson commented on readme/ampathformstranslations.md:
````diff
@@ -0,0 +1,36 @@
+* The `form` attribute must be provided with an existing form name for the translations to load successfully. The translations form resources get names following the following pattern `<form_name>_translations_<locale>`.
+* The UUID must match the identifiers specified in the form's schema, as shown [here](../readme/ampathforms.md) in the translations property, eg:
+```json
+"translations": {
+  "en": ["global-uuid-resource-en", "test-form-en"],
+  "fr": ["global-fr-uuid-resource-fr", "test-form-fr"]
+}
````
Do we really need that even when using the form resource?
I was under the impression that saving the translations as form resources would avoid having to specify the said translation file UUID in the form.

Ruhanga replied:
Correct @rbuisson. Specifying this form translation uuid is not a requirement. To make this non umbigious, let me remove it.

ibacher commented on api/src/main/java/org/openmrs/module/initializer/api/loaders/AmpathFormsTranslationsLoader.java:
```diff
@@ -0,0 +1,76 @@
+			throw new IllegalArgumentException("Uuid is required for AMPATH forms translations loader.");
+		}
+		
+		String jsonTranlsationsFileName = FilenameUtils.removeExtension(file.getName());
```
Do we need this line? The naming convention is for the file-as-stored... I don't necessarily want to impose a naming convention on the file in the configuration folder.

ibacher commented on api/src/main/java/org/openmrs/module/initializer/api/loaders/AmpathFormsTranslationsLoader.java:
```diff
@@ -0,0 +1,76 @@
+		String language = (String) jsonTranslationsDefinition.get("language");
+		if (StringUtils.isBlank(language)) {
+			throw new IllegalArgumentException(
+			        "'language' property is required for AMPATH forms translations loader and should align with locale appended to the file name.");
```
```suggestion
			        "'language' property is required for AMPATH forms translations loader");
```

ibacher commented on api/src/main/java/org/openmrs/module/initializer/api/loaders/AmpathFormsTranslationsLoader.java:
```diff
@@ -0,0 +1,76 @@
+			String formName = (String) jsonTranslationsDefinition.get("form");
+			form = formService.getForm(formName);
+			if (form == null) {
+				throw new IllegalArgumentException("No AMPATH form exists for AMPATH form tranlsations file: "
```
I think we need two checks here:

1. Does the form property exist and have a value
2. Does a form of that name exist

The message here should probably be:

```suggestion
				throw new IllegalArgumentException("Could not find a form named '" + formName + "'. Please ensure a form of this name exists in your configuration.");
```

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class AmpathFormsTranslationsLoaderIntegrationTest extends DomainBaseModuleContextSensitiveTest {
+	
+	@Rule
+    public ExpectedException thrown = ExpectedException.none();
```
Can you make the indentation here consistent (tabs v spaces)

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+		
+		FormResource formResource = formService.getFormResourceByUuid(RESOURCE_UUID);
+		
+		// Verify clob
```
```suggestion
		// Verify
```

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+		// Replay
+		ampathFormsTranslationsLoader.load();
+		
+		FormResource formResource = formService.getFormResourceByUuid(RESOURCE_UUID);
```
This belongs in the Verify section, I think.

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+		
+		ObjectMapper mapper = new ObjectMapper();
+		JsonNode actualObj = mapper.readTree((String) formResource.getValue());
+		Assert.assertEquals("\"c5bf3efe-3798-4052-8dcb-09aacfcbabdc\"", actualObj.get("uuid").toString());
```
Wouldn't it be simpler to just compare the string values and use the `textValue()` on the JsonNode?

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+		
+		FormResource formResource = formService.getFormResourceByUuid(RESOURCE_UUID);
+		
+		// Verify clob
```
```suggestion
		// Verify
```

ibacher commented on api/src/test/java/org/openmrs/module/initializer/api/form/AmpathFormsTranslationsLoaderIntegrationTest.java:
```diff
@@ -0,0 +1,184 @@
+		Assert.assertEquals("\"Autre\"", ampathTranslations.get("translations").get("Other").toString());
+		Assert.assertEquals("\"Enfant\"", ampathTranslations.get("translations").get("Child").toString());
+		
+		String test_file_updated = "src/test/resources/testdata/testAmpathformstranslations/test_form_updated_translations_fr.json";
```
Should have a comment here, e.g. `// Now load updated values`

ibacher commented on api/src/test/resources/testAppDataDir/configuration/ampathformstranslations/test_form_translations_fr.json:
```diff
@@ -0,0 +1,11 @@
+        "Other" : "Autre",
+        "Child" : "Enfant"
+    }
+}
```
```suggestion
}

```

ibacher commented on readme/ampathforms.md:
```diff
@@ -1,70 +1,71 @@
 The JSON source for these forms can be obtained from [AMPATH Form Builder](https://openmrs-spa.org/formbuilder/) by downloading the form, although it can also be written by hand.
 
-**Note** The [AMPATH Form Builder](https://openmrs-spa.org/formbuilder/) does not include a reference to the encounter type by default. For encounter-based forms, you must ensure that you specify an `"encounter"` field with the name of the encounter type for this form or else Initializer will not be able to properly load the form.
+**NOTE:** The [AMPATH Form Builder](https://openmrs-spa.org/formbuilder/) does not include by default a reference to the encounter type. For encounter-based forms, it is important to specify an `"encounter"` field with the name of the encounter type associated with the form, or else Initializer will not be able to properly load the form.
```
I think the original placement of "by default" actually reads better.

```suggestion
**NOTE:** The [AMPATH Form Builder](https://openmrs-spa.org/formbuilder/) does not include a reference to the encounter type by default. For encounter-based forms, it is important to specify an `"encounter"` field with the name of the encounter type associated with the form, or else Initializer will not be able to properly load the form.
```

ibacher commented on readme/ampathformstranslations.md:
````diff
@@ -0,0 +1,29 @@
+```
+
+**NOTE:** 
+* The `form` attribute must be provided with an existing form name for the translations to load successfully. The translations form resources get names following the following pattern `<form_name>_translations_<locale>`.
````
I'd use `<language>` in place of `<locale>` or rename the property to `locale`.

ibacher reviewed (commented):
Thanks @Ruhanga! Functionality-wise, this looks great, just a few comments on polishing this up.

ibacher commented:
Code looks good... Not sure about the build failure...

Ruhanga reviewed (commented):
The fails randomly on with github actions but passes locally @ibacher.

rbuisson commented:
GitHub Actions seems to be failing to get the `org.bahmni.module:bahmnicore-api:jar:0.93-1.1.0` dependency. If removing this dependency from your local Maven repo, you should face the same issue.

rbuisson commented:
Adding this to pom.xml `<repositories>` section shall fix it.
```
        <repository>
            <id>mks-repo</id>
            <name>Mekom Solutions Nexus</name>
            <url>https://nexus.mekomsolutions.net/repository/maven-public</url>
        </repository>
```
